File: web/threejs/src/mobile/MobilePerformanceOptimizer.js
/**
 * MobilePerformanceOptimizer - Enhanced mobile performance optimization for Three.js
 * Implements touch event throttling, adaptive quality scaling, and device-specific optimizations
 */

export class MobilePerformanceOptimizer {
    constructor(options = {}) {
        this.options = {
            touchThrottleInterval: 16, // 60fps instead of 4ms (240fps)
            useRequestAnimationFrame: true,
            enableAdaptiveQuality: true,
            enableGestures: true,
            sensitivity: 1.0,
            targetFPS: 60,
            memoryThresholdMB: 512,
            ...options
        };
        
        // Device capabilities
        this.deviceCapabilities = this.detectDeviceCapabilities();
        
        // Performance monitoring
        this.performanceMetrics = {
            fps: {
                current: 60,
                average: 60,
                min: 60,
                max: 60,
                history: [],
                target: this.options.targetFPS
            },
            memory: {
                used: 0,
                total: 0,
                percentage: 0,
                textures: 0,
                geometries: 0,
                contexts: 0,
                history: []
            },
            quality: {
                current: 'high',
                renderScale: 1.0,
                textureQuality: 'high',
                effectsEnabled: true
            }
        };
        
        // Touch throttling state
        this.touchThrottleState = {
            lastEventTime: {},
            throttledHandlers: new Map(),
            activeTimers: new Map()
        };
        
        // Adaptive quality state
        this.adaptiveQuality = {
            enabled: this.options.enableAdaptiveQuality,
            performanceDropCount: 0,
            dropThreshold: 3,
            lastAdjustment: 0,
            adjustmentCooldown: 2000 // 2 seconds
        };
        
        // Quality presets
        this.qualityPresets = {
            low: {
                renderScale: 0.5,
                textureQuality: 'low',
                enableShadows: false,
                enablePostProcessing: false,
                enableAntialiasing: false,
                maxTextureSize: 1024,
                cullingDistance: 50
            },
            medium: {
                renderScale: 0.75,
                textureQuality: 'medium',
                enableShadows: false,
                enablePostProcessing: true,
                enableAntialiasing: false,
                maxTextureSize: 2048,
                cullingDistance: 100
            },
            high: {
                renderScale: 1.0,
                textureQuality: 'high',
                enableShadows: true,
                enablePostProcessing: true,
                enableAntialiasing: true,
                maxTextureSize: 4096,
                cullingDistance: 200
            }
        };
        
        // Set initial quality based on device
        this.setQualityForDevice();
        
        console.log('📱 MobilePerformanceOptimizer initialized:', {
            device: this.deviceCapabilities,
            quality: this.performanceMetrics.quality.current,
            touchInterval: this.options.touchThrottleInterval
        });
    }
    
    /**
     * Detect device capabilities for optimization
     */
    detectDeviceCapabilities() {
        const userAgent = navigator.userAgent;
        const isMobile = /Mobile|Android|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(userAgent);
        const isTablet = /iPad|Android.*Tablet|Kindle|Silk/i.test(userAgent);
        const isLowEnd = this.detectLowEndDevice();
        
        // Memory estimation
        let memoryGB = 4; // Default assumption
        if (navigator.deviceMemory) {
            memoryGB = navigator.deviceMemory;
        } else if (isMobile) {
            memoryGB = isTablet ? 3 : 2; // Rough mobile estimation
        }
        
        // CPU cores
        const cores = navigator.hardwareConcurrency || 4;
        
        // WebGL support
        const canvas = document.createElement('canvas');
        const webgl2Support = !!canvas.getContext('webgl2');
        const webgl1Support = !!canvas.getContext('webgl') || !!canvas.getContext('experimental-webgl');
        
        // Touch support
        const supportsTouch = 'ontouchstart' in window || navigator.maxTouchPoints > 0;
        
        // Pixel ratio
        const pixelRatio = window.devicePixelRatio || 1;
        
        return {
            isMobile,
            isTablet,
            isLowEnd,
            memoryGB,
            cores,
            webgl2Support,
            webgl1Support,
            supportsTouch,
            pixelRatio,
            platform: this.detectPlatform(),
            browser: this.detectBrowser()
        };
    }
    
    /**
     * Detect if device is low-end
     */
    detectLowEndDevice() {
        // Check for known low-end indicators
        const userAgent = navigator.userAgent;
        const memoryGB = navigator.deviceMemory || 2;
        const cores = navigator.hardwareConcurrency || 2;
        
        // Low memory or few cores
        if (memoryGB < 3 || cores < 4) return true;
        
        // Known low-end devices
        const lowEndPatterns = [
            /Android.*Go/i,
            /Android.*Lite/i,
            /iPhone.*5/i,
            /iPhone.*6/i
        ];
        
        return lowEndPatterns.some(pattern => pattern.test(userAgent));
    }
    
    /**
     * Detect platform
     */
    detectPlatform() {
        const userAgent = navigator.userAgent;
        if (/Windows/i.test(userAgent)) return 'Windows';
        if (/Mac/i.test(userAgent)) return 'macOS';
        if (/Linux/i.test(userAgent)) return 'Linux';
        if (/Android/i.test(userAgent)) return 'Android';
        if (/iOS|iPhone|iPad/i.test(userAgent)) return 'iOS';
        return 'Unknown';
    }
    
    /**
     * Detect browser
     */
    detectBrowser() {
        const userAgent = navigator.userAgent;
        if (/Chrome/i.test(userAgent)) return 'Chrome';
        if (/Firefox/i.test(userAgent)) return 'Firefox';
        if (/Safari/i.test(userAgent) && !/Chrome/i.test(userAgent)) return 'Safari';
        if (/Edge/i.test(userAgent)) return 'Edge';
        return 'Unknown';
    }
    
    /**
     * Set quality preset based on device capabilities
     */
    setQualityForDevice() {
        let quality = 'high';
        
        if (this.deviceCapabilities.isLowEnd) {
            quality = 'low';
        } else if (this.deviceCapabilities.isMobile) {
            quality = this.deviceCapabilities.memoryGB >= 4 ? 'medium' : 'low';
        } else {
            quality = this.deviceCapabilities.memoryGB >= 8 ? 'high' : 'medium';
        }
        
        this.setQualityLevel(quality);
    }
    
    /**
     * Create throttled touch event handler
     * Reduces touch event frequency from 240fps to 60fps
     */
    throttleTouchEvents(eventHandler, customInterval = null) {
        const interval = customInterval || this.options.touchThrottleInterval;
        const handlerId = Math.random().toString(36).substring(2, 9);
        
        const throttledHandler = (event) => {
            const eventType = event.type;
            const now = performance.now();
            const lastTime = this.touchThrottleState.lastEventTime[eventType] || 0;
            
            if (now - lastTime >= interval) {
                this.touchThrottleState.lastEventTime[eventType] = now;
                
                if (this.options.useRequestAnimationFrame) {
                    // Use requestAnimationFrame for smooth visual updates
                    requestAnimationFrame(() => {
                        try {
                            eventHandler(event);
                        } catch (error) {
                            console.error('❌ Error in throttled touch handler:', error);
                        }
                    });
                } else {
                    // Direct execution
                    try {
                        eventHandler(event);
                    } catch (error) {
                        console.error('❌ Error in throttled touch handler:', error);
                    }
                }
            }
        };
        
        // Store handler for cleanup
        this.touchThrottleState.throttledHandlers.set(handlerId, throttledHandler);
        
        return {
            handler: throttledHandler,
            id: handlerId,
            dispose: () => this.disposeTouchHandler(handlerId)
        };
    }
    
    /**
     * Dispose of touch handler
     */
    disposeTouchHandler(handlerId) {
        this.touchThrottleState.throttledHandlers.delete(handlerId);
        
        const timer = this.touchThrottleState.activeTimers.get(handlerId);
        if (timer) {
            clearTimeout(timer);
            this.touchThrottleState.activeTimers.delete(handlerId);
        }
    }
    
    /**
     * Update performance metrics
     */
    updatePerformanceMetrics(fps, memoryUsage = {}) {
        // Update FPS metrics
        this.performanceMetrics.fps.current = fps;
        this.performanceMetrics.fps.history.push(fps);
        
        // Keep history limited
        if (this.performanceMetrics.fps.history.length > 60) {
            this.performanceMetrics.fps.history.shift();
        }
        
        // Calculate statistics
        const fpsHistory = this.performanceMetrics.fps.history;
        if (fpsHistory.length > 0) {
            this.performanceMetrics.fps.average = fpsHistory.reduce((a, b) => a + b) / fpsHistory.length;
            this.performanceMetrics.fps.min = Math.min(...fpsHistory);
            this.performanceMetrics.fps.max = Math.max(...fpsHistory);
        }
        
        // Update memory metrics
        if (memoryUsage.used !== undefined) {
            this.performanceMetrics.memory.used = memoryUsage.used;
            this.performanceMetrics.memory.total = memoryUsage.total || 0;
            this.performanceMetrics.memory.percentage = memoryUsage.total > 0 ? 
                (memoryUsage.used / memoryUsage.total) * 100 : 0;
            this.performanceMetrics.memory.textures = memoryUsage.textures || 0;
            this.performanceMetrics.memory.geometries = memoryUsage.geometries || 0;
            this.performanceMetrics.memory.contexts = memoryUsage.contexts || 0;
            
            this.performanceMetrics.memory.history.push(memoryUsage.used);
            if (this.performanceMetrics.memory.history.length > 60) {
                this.performanceMetrics.memory.history.shift();
            }
        }
        
        // Check for adaptive quality adjustment
        if (this.adaptiveQuality.enabled) {
            this.checkPerformanceAndAdjust();
        }
        
        // Send metrics to Flutter if available
        this.sendMetricsToFlutter();
    }
    
    /**
     * Check performance and automatically adjust quality
     */
    checkPerformanceAndAdjust() {
        const now = performance.now();
        if (now - this.adaptiveQuality.lastAdjustment < this.adaptiveQuality.adjustmentCooldown) {
            return; // Too soon since last adjustment
        }
        
        const fps = this.performanceMetrics.fps;
        const targetFPS = fps.target;
        
        // Check if performance is consistently below target
        if (fps.average < targetFPS * 0.8) { // 80% of target FPS
            this.adaptiveQuality.performanceDropCount++;
            
            if (this.adaptiveQuality.performanceDropCount >= this.adaptiveQuality.dropThreshold) {
                this.reduceQuality();
                this.adaptiveQuality.performanceDropCount = 0;
                this.adaptiveQuality.lastAdjustment = now;
            }
        } else if (fps.average > targetFPS * 0.95) { // 95% of target FPS
            this.adaptiveQuality.performanceDropCount = 0;
            
            // Consider increasing quality if performance is good
            if (fps.average > targetFPS && this.performanceMetrics.memory.used < 400) {
                this.increaseQuality();
                this.adaptiveQuality.lastAdjustment = now;
            }
        }
    }
    
    /**
     * Reduce quality for better performance
     */
    reduceQuality() {
        const currentQuality = this.performanceMetrics.quality.current;
        let newQuality = currentQuality;
        
        if (currentQuality === 'high') {
            newQuality = 'medium';
        } else if (currentQuality === 'medium') {
            newQuality = 'low';
        }
        
        if (newQuality !== currentQuality) {
            this.setQualityLevel(newQuality);
            console.log(`📉 Quality reduced to: ${newQuality}`, {
                averageFPS: this.performanceMetrics.fps.average,
                targetFPS: this.performanceMetrics.fps.target,
                memoryMB: this.performanceMetrics.memory.used
            });
        }
    }
    
    /**
     * Increase quality if performance allows
     */
    increaseQuality() {
        const currentQuality = this.performanceMetrics.quality.current;
        let newQuality = currentQuality;
        
        if (currentQuality === 'low') {
            newQuality = 'medium';
        } else if (currentQuality === 'medium') {
            newQuality = 'high';
        }
        
        if (newQuality !== currentQuality) {
            this.setQualityLevel(newQuality);
            console.log(`📈 Quality increased to: ${newQuality}`, {
                averageFPS: this.performanceMetrics.fps.average,
                targetFPS: this.performanceMetrics.fps.target,
                memoryMB: this.performanceMetrics.memory.used
            });
        }
    }
    
    /**
     * Set quality level
     */
    setQualityLevel(quality) {
        if (!this.qualityPresets[quality]) {
            console.warn(`⚠️ Unknown quality level: ${quality}`);
            return;
        }
        
        const preset = this.qualityPresets[quality];
        this.performanceMetrics.quality = {
            current: quality,
            renderScale: preset.renderScale,
            textureQuality: preset.textureQuality,
            effectsEnabled: preset.enablePostProcessing
        };
        
        console.log(`🎯 Quality set to: ${quality}`, preset);
    }
    
    /**
     * Get device-specific performance settings
     */
    getPerformanceSettings() {
        const settings = {
            ...this.qualityPresets[this.performanceMetrics.quality.current],
            touchThrottleInterval: this.options.touchThrottleInterval,
            useRequestAnimationFrame: this.options.useRequestAnimationFrame,
            sensitivity: this.getTouchSensitivity(),
            enableGestures: this.options.enableGestures
        };
        
        // Mobile-specific optimizations
        if (this.deviceCapabilities.isMobile) {
            settings.enableShadows = false; // Always disable shadows on mobile
            settings.enableLOD = true;
            settings.pixelRatio = this.deviceCapabilities.isLowEnd ? 1.0 : Math.min(this.deviceCapabilities.pixelRatio, 2.0);
        }
        
        // Browser-specific optimizations
        if (this.deviceCapabilities.browser === 'Safari') {
            settings.enableWebGL2 = false; // Safari WebGL2 issues
            settings.precision = 'mediump';
            settings.enableFloatTextures = false;
        }
        
        return settings;
    }
    
    /**
     * Get device-specific touch sensitivity
     */
    getTouchSensitivity() {
        let sensitivity = this.options.sensitivity;
        
        // Adjust for device type
        if (this.deviceCapabilities.isMobile) {
            if (this.deviceCapabilities.isTablet) {
                sensitivity *= 0.8; // Tablets need less sensitivity
            } else {
                sensitivity *= 1.2; // Phones need more sensitivity
            }
        }
        
        // Adjust for pixel ratio
        if (this.deviceCapabilities.pixelRatio > 2.0) {
            sensitivity *= 0.9; // High DPI screens
        }
        
        return sensitivity;
    }
    
    /**
     * Send performance metrics to Flutter
     */
    sendMetricsToFlutter() {
        if (window.parent && window.parent !== window) {
            try {
                window.parent.postMessage({
                    type: 'mobile_performance_update',
                    source: 'threejs',
                    payload: {
                        fps: this.performanceMetrics.fps,
                        memory: this.performanceMetrics.memory,
                        quality: this.performanceMetrics.quality,
                        device: this.deviceCapabilities
                    },
                    timestamp: Date.now()
                }, '*');
            } catch (error) {
                console.warn('⚠️ Failed to send metrics to Flutter:', error);
            }
        }
    }
    
    /**
     * Enable/disable adaptive quality scaling
     */
    setAdaptiveQualityEnabled(enabled) {
        this.adaptiveQuality.enabled = enabled;
        console.log(`🔄 Adaptive quality scaling ${enabled ? 'enabled' : 'disabled'}`);
    }
    
    /**
     * Get performance statistics
     */
    getPerformanceStats() {
        return {
            fps: this.performanceMetrics.fps,
            memory: this.performanceMetrics.memory,
            quality: this.performanceMetrics.quality,
            device: this.deviceCapabilities,
            adaptiveQuality: {
                enabled: this.adaptiveQuality.enabled,
                performanceDropCount: this.adaptiveQuality.performanceDropCount
            }
        };
    }
    
    /**
     * Dispose of optimizer
     */
    dispose() {
        // Clear all throttled handlers
        for (const [handlerId] of this.touchThrottleState.throttledHandlers) {
            this.disposeTouchHandler(handlerId);
        }
        
        // Clear timers
        for (const timer of this.touchThrottleState.activeTimers.values()) {
            clearTimeout(timer);
        }
        
        // Clear state
        this.touchThrottleState.throttledHandlers.clear();
        this.touchThrottleState.activeTimers.clear();
        this.performanceMetrics.fps.history = [];
        this.performanceMetrics.memory.history = [];
        
        console.log('🗑️ MobilePerformanceOptimizer disposed');
    }
}

// Export for global use
window.MobilePerformanceOptimizer = MobilePerformanceOptimizer;

File: lib/core/webgl/mobile_performance_optimizer.dart
import 'dart:async';
import 'dart:html' as html;
import '../logging/app_logger.dart';
import '../platform/platform_utils.dart';

/// Device capability information for performance optimization
class DeviceCapabilities {
  final bool isMobile;
  final bool isTablet;
  final bool isLowEnd;
  final double memoryGB;
  final int cores;
  final String platform;
  final String browser;
  final bool supportsWebGL2;
  final bool supportsTouch;
  final double pixelRatio;
  
  const DeviceCapabilities({
    required this.isMobile,
    required this.isTablet,
    required this.isLowEnd,
    required this.memoryGB,
    required this.cores,
    required this.platform,
    required this.browser,
    required this.supportsWebGL2,
    required this.supportsTouch,
    required this.pixelRatio,
  });
  
  Map<String, dynamic> toMap() {
    return {
      'isMobile': isMobile,
      'isTablet': isTablet,
      'isLowEnd': isLowEnd,
      'memoryGB': memoryGB,
      'cores': cores,
      'platform': platform,
      'browser': browser,
      'supportsWebGL2': supportsWebGL2,
      'supportsTouch': supportsTouch,
      'pixelRatio': pixelRatio,
    };
  }
}

/// Performance preset configurations for different device types
enum PerformancePreset {
  low('Low', 0.5, 30, 1024),
  medium('Medium', 0.75, 45, 2048),
  high('High', 1.0, 60, 4096),
  ultra('Ultra', 1.2, 60, 8192);
  
  const PerformancePreset(this.displayName, this.renderScale, this.targetFPS, this.maxTextureSize);
  
  final String displayName;
  final double renderScale;
  final int targetFPS;
  final int maxTextureSize;
}

/// Touch event throttling configuration
class TouchThrottleConfig {
  final int intervalMs;
  final bool useRequestAnimationFrame;
  final double sensitivity;
  final bool enableGestures;
  
  const TouchThrottleConfig({
    this.intervalMs = 16, // 60fps (16ms intervals)
    this.useRequestAnimationFrame = true,
    this.sensitivity = 1.0,
    this.enableGestures = true,
  });
}

/// Mobile performance optimizer for WebGL applications
/// Implements touch event throttling, device-specific optimizations, and adaptive quality scaling
class MobilePerformanceOptimizer {
  static const String _logComponent = 'MobilePerformanceOptimizer';
  
  /// Touch throttling configuration
  late final TouchThrottleConfig _touchConfig;
  
  /// Device capabilities
  late final DeviceCapabilities _deviceCapabilities;
  
  /// Current performance preset
  PerformancePreset _currentPreset = PerformancePreset.high;
  
  /// Touch event throttling state
  final Map<String, int> _lastTouchEventTime = {};
  final Map<String, Timer?> _touchThrottleTimers = {};
  
  /// Performance monitoring
  final List<double> _fpsHistory = [];
  final List<double> _memoryHistory = [];
  double _currentFPS = 60.0;
  double _currentMemoryMB = 0.0;
  
  /// Adaptive quality scaling
  bool _adaptiveQualityEnabled = true;
  int _performanceDropCount = 0;
  static const int _performanceDropThreshold = 3;
  
  /// Initialize mobile performance optimizer
  MobilePerformanceOptimizer({
    TouchThrottleConfig? touchConfig,
    DeviceCapabilities? deviceCapabilities,
  }) {
    _touchConfig = touchConfig ?? const TouchThrottleConfig();
    _deviceCapabilities = deviceCapabilities ?? _detectDeviceCapabilities();
    
    // Set initial performance preset based on device capabilities
    _currentPreset = _determineOptimalPreset();
    
    AppLogger.info('Mobile performance optimizer initialized',
      component: _logComponent,
      metadata: {
        'preset': _currentPreset.displayName,
        'device': _deviceCapabilities.toMap(),
        'touchInterval': _touchConfig.intervalMs,
      });
  }
  
  /// Detect device capabilities
  DeviceCapabilities _detectDeviceCapabilities() {
    try {
      // Detect mobile/tablet
      final userAgent = html.window.navigator.userAgent;
      final isMobile = RegExp(r'Mobile|Android|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini').hasMatch(userAgent);
      final isTablet = RegExp(r'iPad|Android.*Tablet|Kindle|Silk').hasMatch(userAgent);
      
      // Detect browser
      String browser = 'Unknown';
      if (userAgent.contains('Chrome')) browser = 'Chrome';
      else if (userAgent.contains('Firefox')) browser = 'Firefox';
      else if (userAgent.contains('Safari')) browser = 'Safari';
      else if (userAgent.contains('Edge')) browser = 'Edge';
      
      // Detect platform
      String platform = 'Unknown';
      if (userAgent.contains('Windows')) platform = 'Windows';
      else if (userAgent.contains('Mac')) platform = 'macOS';
      else if (userAgent.contains('Linux')) platform = 'Linux';
      else if (userAgent.contains('Android')) platform = 'Android';
      else if (userAgent.contains('iOS')) platform = 'iOS';
      
      // Estimate memory (rough approximation)
      double memoryGB = 4.0; // Default assumption
      if (html.window.navigator.deviceMemory != null) {
        memoryGB = html.window.navigator.deviceMemory!.toDouble();
      } else {
        // Rough estimation based on device type
        if (isMobile) {
          memoryGB = isTablet ? 3.0 : 2.0;
        } else {
          memoryGB = 8.0;
        }
      }
      
      // Estimate cores
      int cores = html.window.navigator.hardwareConcurrency ?? 4;
      
      // Check WebGL2 support
      final canvas = html.CanvasElement();
      final webgl2Context = canvas.getContext('webgl2');
      final supportsWebGL2 = webgl2Context != null;
      
      // Check touch support
      final supportsTouch = html.window.navigator.maxTouchPoints != null && 
                           html.window.navigator.maxTouchPoints! > 0;
      
      // Get pixel ratio
      final pixelRatio = (html.window.devicePixelRatio ?? 1.0).toDouble();
      
      // Determine if low-end device
      final isLowEnd = (isMobile && memoryGB < 3.0) || 
                       cores < 4 || 
                       !supportsWebGL2;
      
      return DeviceCapabilities(
        isMobile: isMobile,
        isTablet: isTablet,
        isLowEnd: isLowEnd,
        memoryGB: memoryGB,
        cores: cores,
        platform: platform,
        browser: browser,
        supportsWebGL2: supportsWebGL2,
        supportsTouch: supportsTouch,
        pixelRatio: pixelRatio,
      );
      
    } catch (e) {
      AppLogger.error('Failed to detect device capabilities: $e',
        component: _logComponent,
        error: e);
      
      // Return safe defaults
      return const DeviceCapabilities(
        isMobile: false,
        isTablet: false,
        isLowEnd: false,
        memoryGB: 4.0,
        cores: 4,
        platform: 'Unknown',
        browser: 'Unknown',
        supportsWebGL2: false,
        supportsTouch: false,
        pixelRatio: 1.0,
      );
    }
  }
  
  /// Determine optimal performance preset based on device capabilities
  PerformancePreset _determineOptimalPreset() {
    if (_deviceCapabilities.isLowEnd) {
      return PerformancePreset.low;
    } else if (_deviceCapabilities.isMobile) {
      return _deviceCapabilities.memoryGB >= 4.0 ? PerformancePreset.medium : PerformancePreset.low;
    } else {
      return _deviceCapabilities.memoryGB >= 8.0 ? PerformancePreset.high : PerformancePreset.medium;
    }
  }
  
  /// Create throttled touch event handler
  /// Reduces touch event frequency from 240fps to 60fps for better performance
  Function throttleTouchEvents(Function eventHandler, [int? customInterval]) {
    final interval = customInterval ?? _touchConfig.intervalMs;
    
    return (dynamic event) {
      final eventType = event.runtimeType.toString();
      final now = DateTime.now().millisecondsSinceEpoch;
      final lastTime = _lastTouchEventTime[eventType] ?? 0;
      
      if (now - lastTime >= interval) {
        _lastTouchEventTime[eventType] = now;
        
        if (_touchConfig.useRequestAnimationFrame) {
          // Use requestAnimationFrame for smooth visual updates
          html.window.requestAnimationFrame((num time) {
            try {
              eventHandler(event);
            } catch (e) {
              AppLogger.error('Error in throttled touch handler: $e',
                component: _logComponent,
                error: e);
            }
          });
        } else {
          // Direct execution
          try {
            eventHandler(event);
          } catch (e) {
            AppLogger.error('Error in throttled touch handler: $e',
              component: _logComponent,
              error: e);
          }
        }
      }
    };
  }
  
  /// Apply device-specific performance optimizations
  Map<String, dynamic> getPerformanceSettings() {
    final settings = <String, dynamic>{
      'preset': _currentPreset.displayName,
      'renderScale': _currentPreset.renderScale,
      'targetFPS': _currentPreset.targetFPS,
      'maxTextureSize': _currentPreset.maxTextureSize,
      'touchThrottleInterval': _touchConfig.intervalMs,
      'useRequestAnimationFrame': _touchConfig.useRequestAnimationFrame,
      'sensitivity': _touchConfig.sensitivity,
      'enableGestures': _touchConfig.enableGestures,
    };
    
    // Mobile-specific optimizations
    if (_deviceCapabilities.isMobile) {
      settings.addAll({
        'enableShadows': false,
        'enablePostProcessing': !_deviceCapabilities.isLowEnd,
        'enableAntialiasing': !_deviceCapabilities.isLowEnd,
        'enableLOD': true,
        'cullingDistance': _deviceCapabilities.isLowEnd ? 50.0 : 100.0,
        'pixelRatio': _deviceCapabilities.isLowEnd ? 1.0 : _deviceCapabilities.pixelRatio,
      });
    }
    
    // Browser-specific optimizations
    if (_deviceCapabilities.browser == 'Safari') {
      settings.addAll({
        'enableWebGL2': false, // Safari WebGL2 issues
        'precision': 'mediump',
        'enableFloatTextures': false,
      });
    }
    
    return settings;
  }
  
  /// Update performance metrics and trigger adaptive quality scaling
  void updatePerformanceMetrics(double fps, double memoryMB) {
    _currentFPS = fps;
    _currentMemoryMB = memoryMB;
    
    // Add to history (keep last 60 samples)
    _fpsHistory.add(fps);
    _memoryHistory.add(memoryMB);
    
    if (_fpsHistory.length > 60) {
      _fpsHistory.removeAt(0);
    }
    if (_memoryHistory.length > 60) {
      _memoryHistory.removeAt(0);
    }
    
    // Check for performance issues
    if (_adaptiveQualityEnabled) {
      _checkPerformanceAndAdjust();
    }
  }
  
  /// Check performance and automatically adjust quality if needed
  void _checkPerformanceAndAdjust() {
    if (_fpsHistory.length < 10) return; // Need enough samples
    
    final averageFPS = _fpsHistory.reduce((a, b) => a + b) / _fpsHistory.length;
    final targetFPS = _currentPreset.targetFPS.toDouble();
    
    // Check if performance is consistently below target
    if (averageFPS < targetFPS * 0.8) { // 80% of target FPS
      _performanceDropCount++;
      
      if (_performanceDropCount >= _performanceDropThreshold) {
        _reduceQuality();
        _performanceDropCount = 0;
      }
    } else if (averageFPS > targetFPS * 0.95) { // 95% of target FPS
      _performanceDropCount = 0;
      
      // Consider increasing quality if performance is good
      if (_currentPreset != PerformancePreset.ultra && 
          averageFPS > targetFPS && 
          _currentMemoryMB < 400) { // Memory usage is reasonable
        _increaseQuality();
      }
    }
  }
  
  /// Reduce quality preset for better performance
  void _reduceQuality() {
    final currentIndex = PerformancePreset.values.indexOf(_currentPreset);
    if (currentIndex > 0) {
      _currentPreset = PerformancePreset.values[currentIndex - 1];
      
      AppLogger.info('Performance quality reduced to: ${_currentPreset.displayName}',
        component: _logComponent,
        metadata: {
          'averageFPS': _fpsHistory.isNotEmpty ? _fpsHistory.reduce((a, b) => a + b) / _fpsHistory.length : 0,
          'targetFPS': _currentPreset.targetFPS,
          'memoryMB': _currentMemoryMB,
        });
    }
  }
  
  /// Increase quality preset if performance allows
  void _increaseQuality() {
    final currentIndex = PerformancePreset.values.indexOf(_currentPreset);
    if (currentIndex < PerformancePreset.values.length - 1) {
      _currentPreset = PerformancePreset.values[currentIndex + 1];
      
      AppLogger.info('Performance quality increased to: ${_currentPreset.displayName}',
        component: _logComponent,
        metadata: {
          'averageFPS': _fpsHistory.isNotEmpty ? _fpsHistory.reduce((a, b) => a + b) / _fpsHistory.length : 0,
          'targetFPS': _currentPreset.targetFPS,
          'memoryMB': _currentMemoryMB,
        });
    }
  }
  
  /// Set performance preset manually
  void setPerformancePreset(PerformancePreset preset) {
    _currentPreset = preset;
    _performanceDropCount = 0; // Reset counter
    
    AppLogger.info('Performance preset set to: ${preset.displayName}',
      component: _logComponent);
  }
  
  /// Set quality level (compatibility method for WebGL service)
  void setQualityLevel(dynamic qualityLevel) {
    // Convert quality level to performance preset
    PerformancePreset preset;
    if (qualityLevel.toString().toLowerCase().contains('low')) {
      preset = PerformancePreset.low;
    } else if (qualityLevel.toString().toLowerCase().contains('medium')) {
      preset = PerformancePreset.medium;
    } else if (qualityLevel.toString().toLowerCase().contains('high')) {
      preset = PerformancePreset.high;
    } else {
      preset = PerformancePreset.medium; // Default to medium
    }
    
    setPerformancePreset(preset);
  }
  
  /// Enable or disable adaptive quality scaling
  void setAdaptiveQualityEnabled(bool enabled) {
    _adaptiveQualityEnabled = enabled;
    
    AppLogger.info('Adaptive quality scaling ${enabled ? 'enabled' : 'disabled'}',
      component: _logComponent);
  }
  
  /// Get device-specific scaling factors for touch sensitivity
  Map<String, double> getTouchScalingFactors() {
    final factors = <String, double>{
      'sensitivity': _touchConfig.sensitivity,
      'deadzone': 0.1,
      'acceleration': 1.0,
    };
    
    // Adjust for device type
    if (_deviceCapabilities.isMobile) {
      if (_deviceCapabilities.isTablet) {
        factors['sensitivity'] = factors['sensitivity']! * 0.8; // Tablets need less sensitivity
        factors['deadzone'] = 0.15;
      } else {
        factors['sensitivity'] = factors['sensitivity']! * 1.2; // Phones need more sensitivity
        factors['deadzone'] = 0.05;
      }
    }
    
    // Adjust for pixel ratio
    if (_deviceCapabilities.pixelRatio > 2.0) {
      factors['sensitivity'] = factors['sensitivity']! * 0.9; // High DPI screens
    }
    
    return factors;
  }
  
  /// Get performance statistics
  Map<String, dynamic> getPerformanceStats() {
    return {
      'currentFPS': _currentFPS,
      'averageFPS': _fpsHistory.isNotEmpty ? _fpsHistory.reduce((a, b) => a + b) / _fpsHistory.length : 0,
      'minFPS': _fpsHistory.isNotEmpty ? _fpsHistory.reduce((a, b) => a < b ? a : b) : 0,
      'maxFPS': _fpsHistory.isNotEmpty ? _fpsHistory.reduce((a, b) => a > b ? a : b) : 0,
      'currentMemoryMB': _currentMemoryMB,
      'averageMemoryMB': _memoryHistory.isNotEmpty ? _memoryHistory.reduce((a, b) => a + b) / _memoryHistory.length : 0,
      'currentPreset': _currentPreset.displayName,
      'adaptiveQualityEnabled': _adaptiveQualityEnabled,
      'performanceDropCount': _performanceDropCount,
      'deviceCapabilities': _deviceCapabilities.toMap(),
    };
  }
  
  /// Get current device capabilities
  DeviceCapabilities get deviceCapabilities => _deviceCapabilities;
  
  /// Get current performance preset
  PerformancePreset get currentPreset => _currentPreset;
  
  /// Get touch throttle configuration
  TouchThrottleConfig get touchConfig => _touchConfig;
  
  /// Dispose of resources
  void dispose() {
    // Cancel any active timers
    for (final timer in _touchThrottleTimers.values) {
      timer?.cancel();
    }
    _touchThrottleTimers.clear();
    
    // Clear history
    _fpsHistory.clear();
    _memoryHistory.clear();
    _lastTouchEventTime.clear();
    
    AppLogger.info('Mobile performance optimizer disposed', component: _logComponent);
  }
}